sycl/test-e2e/Graph/Threading/begin_end_recording.cpp

```diff
@@ -59,6 +62,34 @@ int main() {
     Threads[i].join();
   }
 
+  // Reference computation
+  queue QueueRef;
+  exp_ext::command_graph GraphRef{Queue.get_context(), Queue.get_device()};
+
+  T *PtrARef = malloc_device<T>(Size, QueueRef);
+  T *PtrBRef = malloc_device<T>(Size, QueueRef);
+  T *PtrCRef = malloc_device<T>(Size, QueueRef);
+
+  QueueRef.copy(DataA.data(), PtrARef, Size);
+  QueueRef.copy(DataB.data(), PtrBRef, Size);
+  QueueRef.copy(DataC.data(), PtrCRef, Size);
+  QueueRef.wait_and_throw();
+
+  for (unsigned i = 0; i < NumThreads; ++i) {
+    queue MyQueue;
+    GraphRef.begin_recording(MyQueue);
+    run_kernels_usm(MyQueue, Size, PtrA, PtrB, PtrC);
+    GraphRef.end_recording(MyQueue);
+  }
+
+  auto GraphImpl = sycl::detail::getSyclObjImpl(Graph);
+  auto GraphRefImpl = sycl::detail::getSyclObjImpl(Graph);
```

Comment:
I don't think testing `getSyclObjImpl()` is something we should be doing in `test-e2e`. "e2e" stands for End To End, and is intended for testing code users can write (no other e2e tests use  `getSyclObjImpl()`) . Users won't have access to `graph_impl.hpp` from if they use a binary release or only have an install directory from a build.

I think `hasSimilarStructure` is important for testing though, but what I think this means is that we need to move the threading tests which use`getSyclObjImpl`  from `test-e2e` to the command-graph unittest - `unittests/Extension/CommandGraph.cpp`.

Reply:
We chatted earlier about the unit tests emulating pi/ur, so any tests which require a device to run on not being suitable for moving. It looks like most of these threading tests don't actually submit an executable graph. `submit.cpp` does, but we're not checking the graph structure there anyway. `finalize.cpp` does too, but I think we could modify it to remove the `handler::ext_oneapi_graph()` call.

So I think it's worth trying to port these tests to the unittest framework and seeing how that goes.

Reply:
I moved threading tests that don't require a device to run to unitests.
However, I prefer to keep "finalize.cpp" test in test-e2e to keep looking for memory leak on this test. 
A thread access issue was only revealed by the memory leak check of this test.